File: kotlin/975-range-sum-of-bst-Easy/Solution.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun rangeSumBST(root: TreeNode?, low: Int, high: Int): Int {
        with(root) {
            return when {
                this == null -> 0 
                `val` < low -> rangeSumBST(right, low, high)
                `val` > high -> rangeSumBST(left, low, high)
                else -> rangeSumBST(left, low, high) + `val` + rangeSumBST(right, low, high)
            }
        }
    }
}

File: scala/975-range-sum-of-bst-Easy/Solution.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(var _value: Int) {
 *   var value: Int = _value
 *   var left: TreeNode = null
 *   var right: TreeNode = null
 * }
 */
object Solution {
    def rangeSumBST(root: TreeNode, L: Int, R: Int): Int = {
        def rangeSum(node: TreeNode): Int = {
            if (node == null) 
                0
            else if (node.value < L)
                rangeSum(node.right)
            else if (node.value > R)
                rangeSum(node.left)
            else
                rangeSum(node.left) + node.value + rangeSum(node.right)
        }
        rangeSum(root)
    }
}
